posthog/clickhouse/client/limit.py

```diff
@@ -0,0 +1,64 @@
+from functools import wraps
+from typing import Optional
+from collections.abc import Callable
+
+from celery import current_task
+from prometheus_client import Counter
+
+from posthog import redis
+
+# Lua script for atomic increment and check
+lua_script = """
+local current_count = redis.call('GET', KEYS[1])
+if current_count and tonumber(current_count) >= tonumber(ARGV[1]) then
+    return 0
+else
+    redis.call('INCR', KEYS[1])
+    return 1
+end
+"""
+
+CONCURRENT_TASKS_LIMIT_EXCEEDED_COUNTER = Counter(
+    "posthog_celery_task_concurrency_limit_exceeded",
+    "Number of times a Celery task exceeded the concurrency limit",
+    ["task_name", "limit", "key"],
+)
+
+
+class CeleryConcurrencyLimitExceeded(Exception):
+    pass
+
+
+def limit_concurrency(max_concurrent_tasks: int, key: Optional[Callable] = None) -> Callable:
+    def decorator(task_func):
+        @wraps(task_func)
+        def wrapper(*args, **kwargs):
+            task_name = current_task.name
+            redis_client = redis.get_client()
+            running_tasks_key = f"celery_running_tasks:{task_name}"
+            if key:
+                dynamic_key = key(*args, **kwargs)
+                running_tasks_key = f"{running_tasks_key}:{dynamic_key}"
+            else:
+                dynamic_key = None
+
+            # Atomically check and increment running task count
+            if redis_client.eval(lua_script, 1, running_tasks_key, max_concurrent_tasks) == 0:
+                CONCURRENT_TASKS_LIMIT_EXCEEDED_COUNTER.labels(
+                    task_name=task_name, limit=max_concurrent_tasks, key=dynamic_key
+                ).inc()
+
+                raise CeleryConcurrencyLimitExceeded(
+                    f"Exceeded maximum concurrent tasks limit: {max_concurrent_tasks} for key: {dynamic_key}"
+                )
+
+            try:
+                # Execute the task
+                return task_func(*args, **kwargs)
+            finally:
```

Comment:
I'm not sure what our failure rate is, but just recently (a couple weeks ago) there were actors queries that were just blowing up and getting OOM Killed by kubernetes. These will not be allowed to call finally.

Every failure will increment on one these over time, and if a team has enough, they'll stop being able to run queries I think.

We probably need to spin a task on a timer that pushes these ref counters down over time (like pushing it down once every 10 minutes is okay) to account for any errors that don't get to decrement, otherwise users will run eventually run into a place where they can't run any queries.

Reply:
That's a good point.. if something crashes then who decreases the counter.

Reply:
Added a sorted set. When the limit is hit, we purge expired tasks. Meaning even if it goes off for some reason, but we never hit the limit, there is also no overhead of checking.